### PR TITLE
Sdba add loess

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [ ] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #xyz
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [ ] HISTORY.rst has been updated (with summary of main changes)
- [ ] `bumpversion (minor / major / patch)` has been called on this branch
- [ ] Tags have been pushed (`git push --tags`)

* **What kind of change does this PR introduce?** 
This add the `xclim.sdba.loess` submodule and the `xclim.sdba.detrending.LoessDetrend` object to perform LOESS smoothing and use is in detrending. I based the implementation on Cleveland (1979) and matlab code from P Grenier. 

The LOESS smoothing has the advantage that its stays within the limits of the data itself. In opposition to polynomial regression, a LOESS-detrended precipitation timeseries will never be negative, for example. However, it has really strong border effects and should always be used with extra data on the sides. The size of the "local" window is controlled by parameter `f` ( a kind of percentage of the total domain) and the outermost f * N points should always be considered with suspicion.  (N is the total number of elements in the smoothed array).

I'm not sure xclim is the best place for such a general smoothing procedure, but it doesn't exist in scipy or numpy... And I didn't want to add a under-maintained dependency. So here we are.

* **Does this PR introduce a breaking change?** 
No

* **Other information**:
This might interest @RondeauG,

### CODE REVIEW

File: xclim/sdba/loess.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
+"""LOESS smoothing module"""
```
Comment (Zeitsperre) [inline xclim/sdba/loess.py:1]:
```suggestion
"""LOESS smoothing module."""
```

File: xclim/sdba/loess.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+import xarray as xr
+
+
+@numba.njit
+def _gaussian_weighting(x):  # pragma: no cover
+    """
+    Kernel function for loess with a gaussian shape.
+    The span f covers 95% of the gaussian.
+    """
+    w = np.exp(-(x ** 2) / (2 * (1 / 1.96) ** 2))
```
Comment (Zeitsperre) [inline xclim/sdba/loess.py:15]:
```suggestion
    """
    
    w = np.exp(-(x ** 2) / (2 * (1 / 1.96) ** 2))
```

File: xclim/sdba/loess.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+import numba
+import numpy as np
+import xarray as xr
+
+
+@numba.njit
+def _gaussian_weighting(x):  # pragma: no cover
+    """
+    Kernel function for loess with a gaussian shape.
+    The span f covers 95% of the gaussian.
```
Comment (Zeitsperre) [inline xclim/sdba/loess.py:13]:
```suggestion
    Kernel function for loess with a gaussian shape.
    
    The span f covers 95% of the gaussian.
```

File: xclim/sdba/detrending.py [hunk 4]
```diff
@@ -160,3 +159,70 @@ def _get_trend(self, grpd, dim="time"):
...
         return trend
+
+
+class LoessDetrend(BaseDetrend):
+    """
+    Detrend time series using a LOESS regression.
+
+    The fit is a piecewise linear regression. For each point, the contribution of all
+    neighbors is weighted by a bell-shaped curve (gaussian) with parameters sigma (std).
+    The x-coordinate of the dataarray is normalized to 0..1 before the regression is computed.
```
Comment (huard) [inline xclim/sdba/detrending.py:170]:
```suggestion
    The x-coordinate of the dataarray is scaled to [0,1] before the regression is computed.
```

File: xclim/sdba/detrending.py [hunk 5]
```diff
@@ -160,3 +159,70 @@ def _get_trend(self, grpd, dim="time"):
...
+    weights : ["tricube", "gaussian"]
+      Shape of the weighting function:
+      "tricube" : a smooth top-hat like curve, f gives the span of non-zero values.
+      "gaussian" : a gaussian curve, f gives the span for 95% of the values.
+
+    Notes
+    -----
+    LOESS smoothing is computationally expensive. As it relies on a loop on gridpoints, it
+    can be useful to use smaller than usual chunks.
+    Moreover, it suffers from heavy boundary effects. As a thumb rule, the outermost f * N points
```
Comment (huard) [inline xclim/sdba/detrending.py:194]:
```suggestion
    Moreover, it suffers from heavy boundary effects. As a rule of thumb, the outermost f * N points
```

File: xclim/sdba/detrending.py [hunk 6]
```diff
@@ -160,3 +159,70 @@ def _get_trend(self, grpd, dim="time"):
...
+            f=self.f,
+            niter=self.niter,
+            d=self.d,
+            weights=self.weights,
+        )
+        trend.name = "trend"
+        return trend.to_dataset()
+
+    def get_trend(self, da: xr.DataArray):
+        """Get the trend computed from the fit, the fitting dim as found on da.
```
Comment (huard) [inline xclim/sdba/detrending.py:216]: Clarify docstring. Not sure what "fitting dim" refers to.

File: xclim/sdba/loess.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    dim: str = "time",
+    d: int = 1,
+    f: float = 0.5,
+    niter: int = 2,
+    weights: Union[str, callable] = "tricube",
+):
+    r"""Locally weighted regression in 1D: fits a nonparametric regression curve to a scatterplot.
+
+    Returns a smoothed curve along given dimension. The regression is computed for each point using
+    a subset of neigboring points as given from evaluation the weighting function locally.
```
Comment (huard) [inline xclim/sdba/loess.py:107]:
```suggestion
    a subset of neigboring points as given from evaluating the weighting function locally.
```

**Comment by aulemahal** [comment]
@huard I pushed the new notebook directly in this PR since it is not going to be merged very soon.

**Comment by huard** [comment]
In the notebook, you say "As it can already be seen in this small example, LOESS smoothing suffers from heavy boundary effects.", but it's not clear what you are referring to. The fit in black looks good in the figure.

**Comment by huard** [comment]
Tried to use `axvspan` to identify the domain that should be discarded in the last graphic, but I get an error from matplotlib.

**Comment by aulemahal** [comment]
@huard  I have modified the notebook to better explain this. Sadly, the `axvspan` bug comes from  `nc-time-axis`. (See SciTools/nc-time-axis#47).

**Comment by huard** [comment]
This is good to go for me.

**Pull request merged**